# Singular Value Decomposition (SVD)

The __singular value decomposition__ (SVD) is a highly relevant method in data science and machine learning, and is where we can bring together all we had learned in previous sections about vectors and matrices. SVD is a method that allows us to decompose a rectangular matrix into a different form. It is tailored to a given problem, as it depends on the data inside a matrix for each instance. Given that we can represent the data in a new form, it is useful for data reduction, and even the basis of __principal component analysis (PCA)__, which we will cover in a later section.

SVD is regarded as one of the most important techniques in linear algebra. It is used in the Google rank algorithm to find the best matches for your search, recommender systems like Netflix and Amazon, and even facial recognition with a SVD representation fo human faces. It is also scalable to any size data set, hence why even Google can do it!!

We can apply this method to any rectangular matrix. If we consider a MxN rectangular matrix, A, SVD allows us to decompose it in the following form:

$$A = U\Sigma V^{T} = 
\begin{bmatrix}
| & | &   & | \\ \mathbf{\vec{u_{1}}} & \mathbf{\vec{u_{2}}} & ... & \mathbf{\vec{u_{N}}} \\ | & | &   & | \end{bmatrix}
\begin{bmatrix}
\sigma_{1} & 0 & ... & 0 \\ 0 & \sigma_{2} &  & 0 \\ \vdots &  & \ddots & \vdots \\ 0 & ... & ... & \sigma_{M} \\ -- & -- & -- & -- \\ 0 & 0 & ... & 0 \\ \vdots & \vdots &  & \vdots \\ 0 & 0 & ... & 0 
\end{bmatrix}
\begin{bmatrix}
| & | &   & | \\ \mathbf{\vec{v_{1}}} & \mathbf{\vec{v_{2}}} & ... & \mathbf{\vec{v_{N}}} \\ | & | &   & | 
\end{bmatrix} ^{T}
$$

Where:
- __U__ is a NxN square, orthogonal matrix, known as the __right singular vectors__
- __V__ is a MxM square, orthogonal matrix, known as the __left singular vectors__
- $\mathbf{\Sigma}$ is a MxN diagonal matrix, containing the __singular values__ along its diagonal

These three matrices are unique for any given rectangular matrix (except by flipping the sign of each vector). So how do we carry out SVD? This is where our understanding of eigenvalues and eigenvectors comes in, where construct each of the three matrices individually:
- U: composed of the the orthonormalized eigenvectors of $AA^{T}$
- V: composed of the orthonormalized eigenvectors of $A^{T}A$
- $\Sigma$ is composed of the square root of the corresponding eigenvalues of $AA^{T}$ or $A^{T}A$

An interesting property is that the eigenvalues of $AA^{T}$ and $A^{T}A$ are always the same so we can pick either one of them. Let us now compute the SVD of the simple matrix below. We will first use NumPy to carry out the invidiual steps, the confirm our answers with the in-built SVD function. 

## Function to sort eigenvalues in descending order
## Use my example, but show second example to show how order is not guaranteed with np.linalg.eig


We will start by determining the matrix U:

$$A = \begin{bmatrix} 1 & 3 & 2 \\ 2 & 2 & 1 \end{bmatrix} $$

We can use NumPy to compute $AA^{T}$:

In [1]:
def reorder_eig_vectors(eig_tuple):
    eig_values, eig_vectors = eig_tuple
    idx = eig_values.argsort()[::-1]
    eig_values = eig_values[idx]
    eig_vectors = eig_vectors[:, idx]
    return eig_values, eig_vectors

In [7]:
import numpy as np

# Define our original matrix
A = np.array([[1,3,2],
              [2,2,1]])
# A = np.array([[3,1,1],
#               [-1,3,1]])
A_T = np.transpose(A)

AA_T = np.matmul(A,A_T)
print(AA_T)

[[14 10]
 [10  9]]


Now that we know that:
$$ AA^{T} = \begin{bmatrix} 14 & 10 \\ 10 & 9 \end{bmatrix}$$

We can compute its eigenvectors and eigenvalues using NumPy:

In [8]:
U_ = np.linalg.eig(AA_T)
U_eigvalues,U_eigvectors = reorder_eig_vectors(U_)
print("Eigenvalues:",U_eigvalues)
print("Eigenvectors:",U_eigvectors)

Eigenvalues: [21.80776406  1.19223594]
Eigenvectors: [[ 0.78820544 -0.61541221]
 [ 0.61541221  0.78820544]]


We then need to carry out orthonormalization to make all the columns of our matrix orthonormal with respect to each other. We will use the previously encountered Gram-Schmidt orthonormalization process, as shown below:  

In [9]:
U,_ = np.linalg.qr(U_eigvectors)
print("U =",U)

U = [[-0.78820544 -0.61541221]
 [-0.61541221  0.78820544]]


Now we have our first matrix, U:
$$U = \begin{bmatrix} -0.788 & -0.615 \\ -0.615 & 0.788 \end{bmatrix}$$

The same procedure can be applied to determine the matrix, $V^{T}$, as shown below:

In [10]:
# Computing the augmented matrix
A_TA = np.matmul(A_T,A)
print(A_TA)
print()

# Computing eigenvalues and eigenvectors
V_ = np.linalg.eig(A_TA)
V_eigvalues,V_eigvectors = reorder_eig_vectors(V_)
V,_ = np.linalg.qr(V_eigvectors)
V_T = V.T
print("Eigenvalues:",V_eigvalues)
print("Eigenvectors:",V_eigvectors)
print()
print(V)
print()
print("V_T =",V_T)
print()

[[ 5  7  4]
 [ 7 13  8]
 [ 4  8  5]]

Eigenvalues: [2.18077641e+01 1.19223594e+00 1.21630936e-15]
Eigenvectors: [[ 0.4323517   0.88011958  0.19611614]
 [ 0.76992171 -0.24711681 -0.58834841]
 [ 0.46935336 -0.4053675   0.78446454]]

[[-0.4323517   0.88011958  0.19611614]
 [-0.76992171 -0.24711681 -0.58834841]
 [-0.46935336 -0.4053675   0.78446454]]

V_T = [[-0.4323517  -0.76992171 -0.46935336]
 [ 0.88011958 -0.24711681 -0.4053675 ]
 [ 0.19611614 -0.58834841  0.78446454]]



We can now construct $V^{T}$ from the above eigenvectors, and $\Sigma$ from the above eigenvalues:
$$ V = \begin{bmatrix} -0.432 & 0.880 & 0.196 \\ -0.769 & -0.247 & -0.588 \\ -0.469 & -0.405 & 0.784 \end{bmatrix}, \;
V^{T} = \begin{bmatrix} -0.432 & -0.769 & -0.469 \\ 0.880 & -0.247 & -0.405 \\ 0.196 & -0.588 & 0.784 \end{bmatrix}$$
$$ \Sigma = \begin{bmatrix} \sqrt{21.808} & 0 & 0 \\0 & \sqrt{1.192} & 0 \end{bmatrix} = 
\begin{bmatrix} 4.670 & 0 & 0 \\0 & 1.092 & 0 \end{bmatrix}$$

We can even confirm our result, as the matrix product of the three matrices we have constructed should simply return A, and the in-built SVD function should return the same three matrices.

In [11]:
# Declaring our Sigma matrix (U and V have already been declared)
rows = len(U)
columns = len(V_T)
Sigma = np.zeros((rows,columns))
np.fill_diagonal(Sigma,np.sqrt(U_eigvalues))
print("Sigma = ",Sigma)
print(U)
print(V_T)
print()

# Numpy function allclose to check if it is the same as the original

# Checking whether we get the original matrix
A_check = np.matmul(U,np.matmul(Sigma,V_T))
print(A_check)
print()


# Checking whether we get the same matrices with the in-built SVD function
U,S,V_T = np.linalg.svd(A)
print(U)
print(S)
print(V_T)

Sigma =  [[4.66987838 0.         0.        ]
 [0.         1.09189557 0.        ]]
[[-0.78820544 -0.61541221]
 [-0.61541221  0.78820544]]
[[-0.4323517  -0.76992171 -0.46935336]
 [ 0.88011958 -0.24711681 -0.4053675 ]
 [ 0.19611614 -0.58834841  0.78446454]]

[[1. 3. 2.]
 [2. 2. 1.]]

[[-0.78820544 -0.61541221]
 [-0.61541221  0.78820544]]
[4.66987838 1.09189557]
[[-0.4323517  -0.76992171 -0.46935336]
 [ 0.88011958 -0.24711681 -0.4053675 ]
 [ 0.19611614 -0.58834841  0.78446454]]


This decomposition contains all the pieces that we need to put our original matrix back together. The singular values($\sigma_{i}$) are arranged in a hierarchical way such that:
$$\sigma_{1} > \sigma_{2} > ... > \sigma_{M}$$

Since diagonal matrices scale each row of a matrix by the respective diagonal component, when we carry out SVD, __these singular values represent the importance of the vectors in U and V__. Given this structure, the vectors in U and V are also arranged according to their importance. This allows us to carry out what is known as the __reduced singular value decomposition__. By only keeping the first R components of the SVD representation, where R < M, R < N, given that R is large enough, we can obtain an __approximation__ of the original matrix, $\hat{A}$, that is good enough given our application of SVD with a smaller number of coefficients.

# (MENTION IMPORTANCE OF REDUCING DIMENSIONALITY)
# use matrix example where SVD actually helps reducing dimensionality

# Extracting decomposed matrices for approximation placed into a function, maybe use it as well for image reconstruction?

# Introduce diag() in explanation for sigma matrix and for the linear algebra section


Take the example below, we have a 10x10 matrix:

In [11]:
# Defining our matrix
A = np.random.randint(0,10,(10,10))
print(A)

[[5 7 3 0 4 0 9 7 9 6]
 [9 5 2 5 5 7 9 8 6 5]
 [4 4 1 7 5 4 9 9 5 8]
 [3 7 2 3 8 8 5 6 3 6]
 [3 2 7 0 5 8 8 2 0 3]
 [0 7 9 0 6 1 3 4 4 3]
 [9 9 7 9 2 4 8 5 3 8]
 [6 1 4 1 0 5 4 7 3 7]
 [3 8 0 6 0 5 3 0 1 7]
 [5 0 1 2 3 3 4 9 3 9]]


In [13]:
# Computing the SVD 
U,S,V_T = np.linalg.svd(A)

# Number of components we wanna keep (R)
R = 8

# Only keeping the first R components
U_hat = U[:,0:R]
S_hat = S[0:R]
V_T_hat = V_T[0:R,:]

# Constructing the diagonal matrix
sigma_hat = np.zeros((R,R))
for i in range(R):
    sigma_hat[i,i] = S_hat[i]

# Computing our original matrix approximation from the reduced SVD
A_hat = np.matmul(U_hat,np.matmul(sigma_hat,V_T_hat))
print(A_hat)

[[ 4.84884254  6.96179313  3.05949426  0.08877473  3.85575737  0.13587008
   8.9728408   7.03902666  9.18481535  5.93568023]
 [ 9.10119284  5.08597371  1.93859805  4.90357869  5.06917513  6.89090344
   9.05011271  8.04615316  5.78193968  5.01409054]
 [ 4.38014159  4.00539342  0.88277458  6.83228916  5.40387891  3.6855405
   9.02035378  8.79331625  4.67686882  8.20525573]
 [ 2.84635238  6.90319229  2.08118164  3.12574301  7.87967018  8.15551816
   4.94174434  5.97029154  3.27840804  5.96242623]
 [ 2.87977936  2.00787912  7.03364913  0.04716852  4.86792583  8.09657006
   7.99863048  2.07683494  0.08722005  2.93049012]
 [ 0.19950883  7.07214519  8.91371783 -0.13047245  6.18053386  0.81414664
   3.0473283   3.97447973  3.7221464   3.07447763]
 [ 8.71582994  8.93174805  7.11056971  9.16470312  1.72720815  4.25435662
   7.95083208  5.07764754  3.3418609   7.87736643]
 [ 6.49348084  0.97371235  3.85971462  0.80267477  0.53939149  4.60178164
   4.00882256  6.6918543   2.63252294  7.28241964]
 

As we can see, for the simple case above, only one single component was enough to fully restore the original matrix. This is what makes SVD such a powerful technique in data reduction. Another example below is in image compression, where we can use SVD to compress an image into a much smaller amount of information.
# ADD IMAGE COMPRESSION SVD EXAMPLE FOR A NICE VISUAL UNDERSTANDING